upup/pkg/fi/cloudup/dotasks/loadbalancer.go

```diff
@@ -0,0 +1,188 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dotasks
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/digitalocean/godo"
+
+	"k8s.io/klog"
+	"k8s.io/kops/pkg/resources/digitalocean"
+	"k8s.io/kops/upup/pkg/fi"
+	"k8s.io/kops/upup/pkg/fi/cloudup/do"
+)
+
+//go:generate fitask -type=LoadBalancer
+type LoadBalancer struct {
+	Name      *string
+	ID        *string
+	Lifecycle *fi.Lifecycle
+
+	Region     *string
+	DropletTag *string
+	IPAddress  *string
+}
+
+var _ fi.CompareWithID = &LoadBalancer{}
+
+func (lb *LoadBalancer) CompareWithID() *string {
+	return lb.ID
+}
+
+func (lb *LoadBalancer) Find(c *fi.Context) (*LoadBalancer, error) {
+	cloud := c.Cloud.(*digitalocean.Cloud)
+	lbService := cloud.LoadBalancers()
+
+	if fi.StringValue(lb.ID) != "" {
+		loadbalancer, _, err := lbService.Get(context.TODO(), fi.StringValue(lb.ID))
+
+		if err != nil {
+			return nil, fmt.Errorf("load balancer service get request returned error %v", err)
+		}
+
+		// do another check to double-sure we are talking to the right load balancer.
+		return &LoadBalancer{
+			Name:      fi.String(loadbalancer.Name),
+			ID:        fi.String(loadbalancer.ID),
+			Lifecycle: lb.Lifecycle,
+			Region:    fi.String(loadbalancer.Region.Slug),
+		}, nil
+	}
+
+	// Loadbalancer = nil if not found
+	return nil, nil
```

Comment:
nit: if we do

```go
if fi.StringValue(lb.ID) == "" {
  // Loadbalancer = nil if not found
  return nil, nil
}

cloud := c.Cloud.(*digitalocean.Cloud)
lbService := cloud.LoadBalancers()
loadbalancer, _, err := lbService.Get(context.TODO(), fi.StringValue(lb.ID))
[...]
```

then we do not need to have the intended if-branch for the case where the LB ID exists. The code also seems easier to read because the check for empty LB ID and the action we take are not so far apart anymore.

(I also moved the `cloud` and `lbService` variables below since they are not needed unless we have an ID.)

Reply:
sure, that's a good suggestion. Updated accordingly.